code/items/weapons/NewtonLauncher.cs

```diff
@@ -0,0 +1,135 @@
+    {
+        public override string ViewModelPath => "weapons/rust_pistol/v_rust_pistol.vmdl";
+        public override SlotType SlotType => SlotType.Equipment;
+        public override bool UnlimitedAmmo => true;
+        public override int ClipSize => 1;
+        public override float PrimaryRate => 1f;
+        public override float SecondaryRate => 1.0f;
+        public override float ReloadTime => 2.3f;
+        public override float DeployTime => 0.4f;
+        public override int BaseDamage => 3;
+
+        public virtual int Price => 100;
+
+        public static int NEWTON_FORCE_MIN => 300;
+        public static int NEWTON_FORCE_MAX => 700;
+        public static float NEWTON_CHARGE_TIME => 3;
+
+        public bool IsCharging { get; set; } = false;
+        public float ChargingStartTime;
+
+        public override void Spawn()
+        {
+            base.Spawn();
+
+            SetModel("weapons/rust_pistol/rust_pistol.vmdl");
+        }
+
+        public override void OnActive()
+        {
+            base.OnActive();
+            IsCharging = false;
+        }
+
+        public override void ActiveEnd(Entity ent, bool dropped)
+        {
+            base.ActiveEnd(ent, dropped);
+            IsCharging = false;
+        }
+
+        public override void AttackPrimary()
+        {
+            if (IsCharging) return;
+
+            IsCharging = true;
+            ChargingStartTime = Time.Now;
+        }
+
+        public void ReleaseCharge()
+        {
+            (Owner as AnimEntity).SetAnimBool("b_attack", true);
+            ShootEffects();
+            PlaySound("rust_pistol.shoot").SetPosition(Position).SetVolume(0.8f);
+            ShootBullet(0f, 0f, BaseDamage, 0f);
+            IsCharging = false;
+            TimeSincePrimaryAttack = 0;
+        }
+
+        public override void ShootBullet(float spread, float force, float damage, float bulletSize)
+        {
+            Vector3 forward = Owner.EyeRot.Forward;
+            forward += (Vector3.Random + Vector3.Random + Vector3.Random + Vector3.Random) * spread * 0.25f;
+            forward = forward.Normal;
+
+            foreach (TraceResult tr in TraceBullet(Owner.EyePos, Owner.EyePos + forward * 5000, bulletSize))
+            {
+                tr.Surface.DoBulletImpact(tr);
+
+                if (!tr.Entity.IsValid() || tr.Entity.IsWorld) return;
+                if (!IsServer)
+                {
+                    continue;
+                }
+
+                using (Prediction.Off())
+                {
+                    DamageInfo damageInfo = DamageInfo.FromBullet(tr.EndPos, forward * 100 * force, damage)
+                        .UsingTraceResult(tr)
+                        .WithAttacker(Owner)
+                        .WithWeapon(this);
+
+                    tr.Entity.TakeDamage(damageInfo);
+
+                    Vector3 pushDirection = new Vector3(-tr.Normal.x, -tr.Normal.y, tr.Normal.z < 0.5F ? 0.5F : -0.5F);
```

Comment:
Could you explain the logic at the end?

Reply:
Terry is a little bit different model than classic terrorist from gmod. If you hit him (for example) in upper part of a hand (or forehead) you expect him to get launched  upwards but tr.Noraml.z can return value (negative number) that would push him to the ground instead. It just reduces this threshold where hitting him will push him in to the ground, while still giving you ability to push him down. It also prevents from Terry's space program (launching him way to high).

It might be simplified to only 0.5F but I thought it will look strange if you hit someone from above and still launching him upwards.

Reply:
Ah yea, gonna try this out as these "magical numbers" seems to be confusing :D But good solution I think.

Reply:
I found better way to do this by changing tr.Normal to tr.Direction.
`Vector3 pushDirection = new Vector3(tr.Direction.x, tr.Direction.y, Math.Min(tr.Direction.z + 0.5F, 0.75F));`